File: 0312-burst-balloons/0312-burst-balloons.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxCoins = function(nums) {
    // 양 끝에 가상의 1 추가
    const arr = [1, ...nums, 1];
    const n = arr.length;
    
    // dp[i][j] = i~j 범위(exclusive)에서 최대 코인
    const dp = Array.from({ length: n }, () => Array(n).fill(0));
    
    // 구간 길이를 늘려가며 계산
    for (let len = 1; len <= nums.length; len++) {
        for (let left = 1; left <= nums.length - len + 1; left++) {
            const right = left + len - 1;
            
            // k를 마지막으로 터뜨리는 경우
            for (let k = left; k <= right; k++) {
                const coins = arr[left - 1] * arr[k] * arr[right + 1];
                const leftCoins = dp[left][k - 1];
                const rightCoins = dp[k + 1][right];
                
                dp[left][right] = Math.max(
                    dp[left][right],
                    leftCoins + coins + rightCoins
                );
            }
        }
    }
    
    return dp[1][nums.length];
};

File: 0312-burst-balloons/0312-burst-balloons.java
class Solution {
    public int maxCoins(int[] nums) {
        int n = nums.length;
        
        // 양 끝에 1 추가
        int[] arr = new int[n + 2];
        arr[0] = arr[n + 1] = 1;
        for (int i = 0; i < n; i++) {
            arr[i + 1] = nums[i];
        }
        
        // dp[i][j] = i~j 범위에서 최대 코인
        int[][] dp = new int[n + 2][n + 2];
        
        // 구간 길이를 늘려가며 계산
        for (int len = 1; len <= n; len++) {
            for (int left = 1; left <= n - len + 1; left++) {
                int right = left + len - 1;
                
                for (int k = left; k <= right; k++) {
                    int coins = arr[left - 1] * arr[k] * arr[right + 1];
                    int total = dp[left][k - 1] + coins + dp[k + 1][right];
                    
                    dp[left][right] = Math.max(dp[left][right], total);
                }
            }
        }
        
        return dp[1][n];
    }
}